From 9a5a0b4f3ee7a48e3a812e9ed37ecf441f63f7da Mon Sep 17 00:00:00 2001
From: Silke Hofstra <silke@slxh.eu>
Date: Wed, 3 Apr 2024 12:23:42 +0200
Subject: [PATCH 2/2] Escape non-printable ASCII characters

Escape characers from ASCII that are not allowed in XML (codepoints 0-8, 11-12, 14-31, 127).
---
 src/iks.c           | 20 ++++++++++++++++++++
 test/tst-iks-utf8.c |  2 ++
 2 files changed, 22 insertions(+)

diff --git a/src/iks.c b/src/iks.c
index ceef571..b7125f1 100644
--- a/src/iks.c
+++ b/src/iks.c
@@ -5,6 +5,8 @@
 */
 
 #include <uchar.h>
+#include <ctype.h>
+#include <stdbool.h>
 #include "common.h"
 #include "iksemel.h"
 
@@ -563,6 +565,11 @@ iks_has_attribs (iks *x)
 }
 
 /*****  Serializing  *****/
+static bool
+is_print(char c)
+{
+	return isprint(c) ||  c == '\t' || c == '\n' || c == '\r';
+}
 
 static size_t
 escape_size (char *src, size_t len)
@@ -585,6 +592,11 @@ escape_size (char *src, size_t len)
 			continue;
 		}
 
+		if (!is_print(c)) {
+			sz += 6;
+			continue;
+		}
+
 		switch (c) {
 			case '&': sz += 5; break;
 			case '\'': sz += 6; break;
@@ -630,6 +642,14 @@ escape (char *dest, char *src, size_t len)
 			continue;
 		}
 
+		if (!is_print(c)) {
+			if (i - j > 0) dest = my_strcat(dest, src + j, i - j);
+			j = i + 1;
+
+			dest = my_strcat(dest, buf, snprintf(buf, sizeof buf, "&#x%02x;", c));
+			continue;
+		}
+
 		if ('&' == c || '<' == c || '>' == c || '\'' == c || '"' == c) {
 			if (i - j > 0) dest = my_strcat (dest, src + j, i - j);
 			j = i + 1;
diff --git a/test/tst-iks-utf8.c b/test/tst-iks-utf8.c
index 1443e6f..f177546 100644
--- a/test/tst-iks-utf8.c
+++ b/test/tst-iks-utf8.c
@@ -18,6 +18,7 @@ int main (int argc, char *argv[])
 						"<emoji>&#x1f4a9;</emoji>"
 						"<invalid></invalid>"
 						"<null></null>"
+						"<nonprint>&#x01;&#x07;&#x0b;&#x7f;</nonprint>"
 						"</test>";
 
 	iks *x = iks_new ("test");
@@ -25,6 +26,7 @@ int main (int argc, char *argv[])
 	iks_insert_cdata (iks_insert (x, "emoji"), "\U0001F4A9", 4);
 	iks_insert_cdata (iks_insert (x, "invalid"), "\x80\x81", 2);
 	iks_insert_cdata (iks_insert (x, "null"), "\0", 1);
+	iks_insert_cdata (iks_insert (x, "nonprint"), "\x1\a\v\x7F", 4);
 
 	char *t = iks_string (iks_stack (x), x);
 	if(!t || strcmp(t, xml) != 0) {
-- 
2.44.0

